Big Red Button

rosieposieeee opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->

![ulRCsvH9Zp](https://github.com/space-wizards/space-station-14/assets/52761126/ac65a0aa-6455-49db-9d72-e17ba1437330)


![8d0yd1PAal](https://github.com/space-wizards/space-station-14/assets/52761126/f3d575f5-c2c3-427c-9212-9baca5d9fe3c)


It's a crime that I didn't think of this last night when I was adding the other buttons. And it's a crime that none of you did either.

Luckily, I woke up this morning with a vision.

Adds a big red button, which is like a regular wall button, but always lockable, and also extremely cool.

Also adds an emergency button which, upon use, also sets the station to alert level Red, and is exclusively Command locked. (The red stripe one is the emergency button.)

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->

It's Cool.

## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->

Uses the lockable button YML from #25631.

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

- [You Fucking Know It Baby] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- add: Added the BIG. RED. BUTTON.

rosieposieeee commented:
i have... no idea... what that test failure means frankly

Kadeo64 commented:
emergency alert imposter among us

MACMAN2003 commented:
Please, Don't Touch Anything.
the tider:

FairlySadPanda commented:
Couple of thoughts as a contributor, feel free to ignore me

1. Shouldn't be specifically an EmergencyButtonComponent/System. "Trigger a change in emergency level on activation" is a broad bit of functionality and components shouldn't assume the entity they're attached to, but be useful when kitbashed.
2. This forces the emission of an empty text message whenever the button has been hit on the announcement panel. This means this interacts with the announcement system, and ergo (sigh) [#24680]. You should add a text field into the component for issuing the message that explains that the big red button has been pressed.

rosieposieeee commented:
> Couple of thoughts as a contributor, feel free to ignore me
> 
>     1. Shouldn't be specifically an EmergencyButtonComponent/System. "Trigger a change in emergency level on activation" is a broad bit of functionality and components shouldn't assume the entity they're attached to, but be useful when kitbashed.
> 
>     2. This forces the emission of an empty text message whenever the button has been hit on the announcement panel. This means this interacts with the announcement system, and ergo (sigh) [[Local/Whisper/Emote/Radio/LOOC/DeadChat Chat Refactor #24680](https://github.com/space-wizards/space-station-14/pull/24680)]. You should add a text field into the component for issuing the message that explains that the big red button has been pressed and this'll have to be another test case for chatfactor to make sure nothing breaks.

will not ignore! thank you for the feedback! ill get to these in a bit.

rosieposieeee commented:
> Couple of thoughts as a contributor, feel free to ignore me
> 
>     1. Shouldn't be specifically an EmergencyButtonComponent/System. "Trigger a change in emergency level on activation" is a broad bit of functionality and components shouldn't assume the entity they're attached to, but be useful when kitbashed.
> 
>     2. This forces the emission of an empty text message whenever the button has been hit on the announcement panel. This means this interacts with the announcement system, and ergo (sigh) [[Local/Whisper/Emote/Radio/LOOC/DeadChat Chat Refactor #24680](https://github.com/space-wizards/space-station-14/pull/24680)]. You should add a text field into the component for issuing the message that explains that the big red button has been pressed and this'll have to be another test case for chatfactor to make sure nothing breaks.

not sure exactly what you want me to do with the text field, but i've added it regardless, and also renamed the component to sound more generic.

0x6273 commented on Content.Server/AlertLevel/ChangeAlertLevelComponent.cs:
```diff
@@ -0,0 +1,18 @@
+using Content.Shared.ChangeAlertLevel;
+using Robust.Shared.Audio;
+
+namespace Content.Server.ChangeAlertLevel
+{
```
Use file-scoped namespace for all new files

rosieposieeee replied:
do you mean like, Content.Server.AlertLevel.ChangeAlertLevel ?

0x6273 replied:
`namespace Content.Server.ChangeAlertLevel;` and get rid of the `{ }`

0x6273 commented on Content.Server/AlertLevel/ChangeAlertLevelComponent.cs:
```diff
@@ -0,0 +1,18 @@
+    [Access(typeof(ChangeAlertLevelSystem))]
+    public sealed partial class ChangeAlertLevelComponent : SharedChangeAlertLevelComponent
+    {
+        [DataField("alertLevelOnActivate")] public string AlertLevelOnActivate = default!;
+        [DataField("textField")] public string TextField = default!;
+        [DataField("clickSound")]
+        public SoundSpecifier ClickSound = new SoundPathSpecifier("/Audio/Machines/button.ogg");
```
- Don't specify DataField names manually. They default to camelCase of the c# field name.
- `[DataField]` and `public string ..` on separate lines
- The first 2 fields should probably be `required: true`

0x6273 commented on Content.Server/AlertLevel/ChangeAlertLevelSystem.cs:
```diff
@@ -0,0 +1,58 @@
+        SubscribeLocalEvent<ChangeAlertLevelComponent, ActivateInWorldEvent>(OnActivated);
+    }
+
+    public void OnActivated(EntityUid uid, ChangeAlertLevelComponent comp, ActivateInWorldEvent args)
```
Use `Entity<TComp>` all for new functions.

rosieposieeee replied:
i'm not entirely clear what this means, can you explain?

0x6273 replied:
https://discord.com/channels/310555209753690112/928056457117696020/1164654158725525575

rosieposieeee replied:
i keep getting a 'method group' error and the only working thing - which visual studio even suggests - seems to be changing back to not using Entity<T>

deltanedas replied:
have to add `ref` before the event type too

rosieposieeee replied:
> have to add `ref` before the event type too

can you go into more detail on this? i cannot for the life of me get any combination of things here to work and i'm having trouble finding any examples in the codebase set up like this to work from, everythings done the old way

rosieposieeee replied:
...nevermind it was because i did the other thing but forgot to update to using the shared version, lol. ref worked

0x6273 commented on Content.Server/AlertLevel/ChangeAlertLevelSystem.cs:
```diff
@@ -0,0 +1,58 @@
+        var buttonXform = Transform(uid);
+        var stationUid = _station.GetStationInMap(buttonXform.MapID);
+        if (stationUid != null)
+            _alertLevel.SetLevel(stationUid.Value, comp.AlertLevelOnActivate, true, true, true, false);
```
Specify the argument names for the bools.

UbaserB commented:
looks oddly round and light-bulby

deltanedas commented on Content.Server/AlertLevel/ChangeAlertLevelComponent.cs:
```diff
@@ -0,0 +1,18 @@
+    /// </summary>
+    [RegisterComponent]
+    [Access(typeof(ChangeAlertLevelSystem))]
+    public sealed partial class ChangeAlertLevelComponent : SharedChangeAlertLevelComponent
```
move this to shared and remove the abstract component, use an abstract empty system for access

i think using abstract components like this is deprecated, and putting it in shared makes predicting it a little easier down the line

rosieposieeee replied:
sure

rosieposieeee replied:
doing this has broken things so terribly

deltanedas replied:
why

deltanedas commented on Resources/Prototypes/Entities/Structures/Wallmounts/switch.yml:
```diff
@@ -509,6 +509,333 @@
 
+# big red button
+- type: entity
+  id: BigRedButton
```
why doesnt this inherit from signal switch

rosieposieeee replied:
because it doesn't go on the wall, it goes on a table

deltanedas replied:
it will still work even if there is no wall under it, and you could still change the 1 or 2 components if not

rosieposieeee replied:
yeah alright

deltanedas commented on Resources/Prototypes/Entities/Structures/Wallmounts/switch.yml:
```diff
@@ -509,6 +509,333 @@
+    canCollide: false
+  - type: SignalSwitch
+  - type: UseDelay
+    delay: 0.5 # prevent light-toggling auto-clickers.
```
le copy pasted comment from signal button

deltanedas commented on Resources/Prototypes/Entities/Structures/Wallmounts/switch.yml:
```diff
@@ -509,6 +509,333 @@
+    access: [["Atmospherics"]]
+
+- type: entity
+  id: BigRedButtonEmergency
```
just make it inherit signal button to remove copypaste, theres uses for also linking it to e.g. shutters

rosieposieeee replied:
wanted to do this, but inheriting signal button on the emergency one specifically overrides the alert system functionality :(

deltanedas replied:
if alert system one doesnt handle the trigger event it should be fine

rosieposieeee replied:
the issue was that they both set args to handled, so actually, if thats not a thing anymore it should work now. will look into it

rosieposieeee commented:
> looks oddly round and light-bulby

yeah the sprite wasn't quite done. i'll show off the current version once i get it compiling

deltanedas commented on Content.Server/AlertLevel/ChangeAlertLevelComponent.cs:
```diff
@@ -1,18 +1,13 @@
+/// </summary>
+[RegisterComponent]
+[Access(typeof(ChangeAlertLevelSystem))]
+public abstract partial class ChangeAlertLevelComponent : SharedChangeAlertLevelComponent
```
remove this server file entirely

rosieposieeee replied:
👍

deltanedas commented on Content.Shared/AlertLevel/ChangeAlertLevel.cs:
```diff
@@ -1,8 +1,17 @@
+using Robust.Shared.Audio;
 using Robust.Shared.Serialization;
 
 namespace Content.Shared.ChangeAlertLevel
```
file scope

rosieposieeee replied:
it needs this for the audio noise :(

deltanedas replied:
someone saying file scope means change this

```cs
namespace Among.Us
{
    public class Sus {}
}
```

to this

```cs
namespace Among.Us;

public class Sus {}
```

its purely cosmetic and has no impact on audio

deltanedas commented on Content.Shared/AlertLevel/ChangeAlertLevel.cs:
```diff
@@ -1,8 +1,17 @@
-    public abstract partial class SharedChangeAlertLevelComponent : Component
+    public sealed partial class SharedChangeAlertLevelComponent : Component
     {
+        [ViewVariables(VVAccess.ReadWrite), DataField(required: true)]
```
remove `ViewVariables(VVAccess.ReadWrite), ` for all these its automatically readwrite

rosieposieeee replied:
👍

deltanedas commented on Content.Shared/AlertLevel/ChangeAlertLevel.cs:
```diff
@@ -1,8 +1,17 @@
 namespace Content.Shared.ChangeAlertLevel
 {
-    public abstract partial class SharedChangeAlertLevelComponent : Component
+    public sealed partial class SharedChangeAlertLevelComponent : Component
```
ChangeAlertLevelComponent

rosieposieeee commented:
![aLWl5u0cJa](https://github.com/space-wizards/space-station-14/assets/52761126/f1994679-4709-4576-b961-cea4b1328844)

upgraded the sprite :)

UbaserB commented:
the vertical sprite is still elongated

deltanedas commented on Content.Server/AlertLevel/ChangeAlertLevelSystem.cs:
```diff
@@ -0,0 +1,63 @@
+
+        _audio.PlayPvs(ent.Comp.ClickSound, ent.Owner, AudioParams.Default.WithVariation(0.125f).WithVolume(8f));
+
+        args.Handled = true;
```
this is the reason you cant use signal button, either that system or this one will handle the interaction event and prevent the other system from working.

make an event say `ButtonPressedEvent` and raise it in the signal button's system and then in here have all this file just handle it like this

```cs
private void OnButtonPressed(...)
{
     if (_station.GetStationInMap(Transform(ent).MapID) is {} station)
         _alertLevel.SetLevel(station, ent.Comp.AlertLevelOnActivate, playsound: true, announce: true, force: false, locked: false);
}
```

also this should probably use GetOwningStation to support multiple stations in the same map, ask sloth

with these changes the component and system can be renamed to `AlertLevelOnPress` which makes it easier to immediately understand what it does compared to `ChangeAlertLevel`

rosieposieeee replied:
alright, will probably get to this tomorrow, thanks for all the help!

rosieposieeee commented:
> the vertical sprite is still elongated

![image](https://github.com/space-wizards/space-station-14/assets/52761126/69bf635e-846e-4b38-bb35-bfd1b45a0176)
 owsabout this

rosieposieeee commented on Content.Server/AlertLevel/ChangeAlertLevelSystem.cs:
```diff
@@ -0,0 +1,63 @@
+
+        if (stationUid != null)
+            _alertLevel.SetLevel(stationUid.Value, ent.Comp.AlertLevelOnActivate, playsound, announce, force, locked);
+        Dirty(ent.Owner, ent.Comp);
```
👍

deltanedas commented on Content.Server/AlertLevel/AlertLevelOnPressSystem.cs:
```diff
@@ -36,27 +37,12 @@ public override void Initialize()
     }
 
-    public void OnActivated(Entity<ChangeAlertLevelComponent> ent, ref ActivateInWorldEvent args)
+    private void OnButtonPressed(Entity<AlertLevelOnPressComponent> ent, ref ActivateInWorldEvent args)
```
still need to actually make ButtonPressedEvent and change it

also rename the component and move to server

rosieposieeee replied:
wait what. i thought i was supposed to delete the server version and move it entirely to shared

0x6273 commented on Content.Server/DeviceLinking/Systems/SignalSwitchSystem.cs:
```diff
@@ -38,8 +38,18 @@ private void OnActivated(EntityUid uid, SignalSwitchComponent comp, ActivateInWo
     }
 }
+
+// Event to inform any other systems that may want to know that the signal switch was pressed
+public sealed class SwitchPressedEvent : EntityEventArgs
```
New events should be record structs raised by ref if possible.

deltanedas replied:
yep
```cs
[ByRefEvent]
public record struct SwitchPressedEvent();
```

0x6273 commented on Content.Server/AlertLevel/AlertLevelOnPressComponent.cs:
```diff
@@ -0,0 +1,18 @@
+    [DataField(required: true)]
+    public string AlertLevelOnActivate = default!;
+
+    [DataField(required: true)]
+    public string TextField = default!;
```
This is never used?

rosieposieeee replied:
fairlysadpanda suggested it above... but i dont really know what to do with it either, so ill remove it

deltanedas commented on Content.Server/AlertLevel/AlertLevelOnPressComponent.cs:
```diff
@@ -0,0 +1,18 @@
+    [DataField(required: true)]
+    public string TextField = default!;
+
+    [DataField]
```
not used either

deltanedas commented on Content.Server/AlertLevel/AlertLevelOnPressComponent.cs:
```diff
@@ -0,0 +1,18 @@
+public sealed partial class AlertLevelOnPressComponent : SharedAlertLevelOnPressComponent
+{
+    [DataField(required: true)]
+    public string AlertLevelOnActivate = default!;
```
`ProtoId<AlertLevelPrototype>` instead of string and string.Empty instead of default!

rosieposieeee replied:
yeah so this is the only thing i can't get working, since i can't access that from shared.

deltanedas commented on Content.Server/AlertLevel/AlertLevelOnPressSystem.cs:
```diff
@@ -0,0 +1,28 @@
+
+public sealed class AlertLevelOnPressSystem : EntitySystem
+{
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
```
remove unused deps and using declarations

deltanedas commented on Content.Server/DeviceLinking/Systems/SignalSwitchSystem.cs:
```diff
@@ -38,8 +38,18 @@ private void OnActivated(EntityUid uid, SignalSwitchComponent comp, ActivateInWo
             _deviceLink.SendSignal(uid, comp.StatusPort, comp.State);
         }
 
+        RaiseLocalEvent(uid, new SwitchPressedEvent());
```
raise by ref

```cs
var ev = new SwitchPressedEvent();
RaiseLocalEvent(uid, ref ev);
```

deltanedas commented on Content.Server/AlertLevel/AlertLevelOnPressComponent.cs:
```diff
@@ -0,0 +1,18 @@
+using Content.Shared.AlertLevelOnPress;
+using Robust.Shared.Audio;
+
+namespace Content.Server.AlertLevelOnPress;
```
move this to shared and remove the empty SharedAlertLevelOnPress

make abstract SharedAlertLevelOnPressSystem for access

rosieposieeee replied:
only issue with this is accessing the ProtoId<AlertLevelPrototype> from shared

deltanedas commented on Resources/Prototypes/Entities/Structures/Wallmounts/switch.yml:
```diff
@@ -509,6 +509,287 @@
+      map: ["enum.LockVisualLayers.Lock"]
+  - type: AlertLevelOnPress
+    alertLevelOnActivate: red
+    textField: An emergency button was pressed. The station has been set to Alert Level Red.
```
unused

rosieposieeee commented:
so; everything except making the string a ProtoId<AlertLevelPrototype> is in and functional, can't figure that one out. the only concern i have left is whether or not to let the emergency button be forced. normally, you can't change the alert level right after changing it on the comms console, there's a cooldown, and that applies to the emergency button too. i could set it to forced, but that may be too powerful/spammy? on the other hand, without it being forced, the button seemingly does nothing for a little bit when it is on cooldown. not sure what the best approach would be.

ficcialfaint commented on Content.Server/DeviceLinking/Systems/SignalSwitchSystem.cs:
```diff
@@ -38,8 +38,15 @@ private void OnActivated(EntityUid uid, SignalSwitchComponent comp, ActivateInWo
 }
+
+[ByRefEvent]
+// Event to inform any other systems that may want to know that the signal switch was pressed
```
Use [XMLDoc](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/), not the regular comments

ficcialfaint commented on Resources/Prototypes/Entities/Structures/Wallmounts/switch.yml:
```diff
@@ -509,6 +509,286 @@
   - type: AccessReader
     access: [["Atmospherics"]]
 
+# big red button
```
```suggestion
```

rosieposieeee commented:
i have no idea what these test failures are about

0x6273 commented on Content.Server/DeviceLinking/Systems/SignalSwitchSystem.cs:
```diff
@@ -38,8 +38,17 @@ private void OnActivated(EntityUid uid, SignalSwitchComponent comp, ActivateInWo
 }
+
+[ByRefEvent]
+/// <summary>
+// Event to inform any other systems that may want to know that the signal switch was pressed
+/// </summary>
```
```suggestion
/// <summary>
/// Event to inform any other systems that may want to know that the signal switch was pressed
/// </summary>
```

rosieposieeee commented:
this is ready to go unless anyone else had any concerns
